Reorganize compile module and remove deprecated functionality

ricardoV94 opened the pull request:
Removing all the deprecated stuff that nobody uses

**Update:** Many of the performance-critical changes were already adopted in spin-off, so this PR is mostly deprecating old stuff and refactoring things

<details> <summary> Old discussion on perf </summary>

I think after this there's not much more fat to trim. We could try to avoid the input/output list thing with linkers that don't need it but we still need to sort them/ recognize kwargs which is pretty much what's left.

The C part of the call takes 100ns, so the "overhead" in this case is 200-250ns.
For reference, calling `numpy.zeros(100)` takes like ~250ns, and a python identity function is ~20ns.

## Results of the new benchmark
<details>
<summary>Before</summary>

```
--------------------------------------------------------------------------------------------------------------------- benchmark: 4 tests ---------------------------------------------------------------------------------------------------------------------
Name (time in ns)                                                                Min                    Max                  Mean              StdDev                Median                IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_identity_function_overhead_benchmark[cvm_nogc-trust_input=True]        524.0000 (1.0)      55,453.8000 (8.24)       574.7866 (1.0)      326.9482 (3.11)       560.5500 (1.0)      24.5500 (1.0)      301;2121    1,739.7760 (1.0)       82217          20
test_identity_function_overhead_benchmark[cvm-trust_input=True]             590.9999 (1.13)      6,733.0000 (1.0)        654.5138 (1.14)     105.2706 (1.0)        650.9999 (1.16)     30.9999 (1.26)       89;486    1,527.8516 (0.88)      19552           1
test_identity_function_overhead_benchmark[cvm_nogc-trust_input=False]     1,432.0001 (2.73)     31,008.0000 (4.61)     1,608.2195 (2.80)     296.4835 (2.82)     1,583.0001 (2.82)     50.0002 (2.04)     722;1368      621.8057 (0.36)      37638           1
test_identity_function_overhead_benchmark[cvm-trust_input=False]          1,463.0000 (2.79)     11,541.9998 (1.71)     1,614.7948 (2.81)     329.0091 (3.13)     1,582.9999 (2.82)     60.0000 (2.44)       78;152      619.2737 (0.36)       3899           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```
</summary> 
</details>
<details>
<summary>After</summary>

```
--------------------------------------------------------------------------------------------------------------------- benchmark: 4 tests ---------------------------------------------------------------------------------------------------------------------
Name (time in ns)                                                                Min                    Max                  Mean              StdDev                Median                IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_identity_function_overhead_benchmark[cvm_nogc-trust_input=True]        322.7857 (1.0)       5,672.7143 (1.0)        358.2689 (1.0)       75.7251 (1.0)        354.9286 (1.0)      12.2143 (1.0)     1548;2274    2,791.1995 (1.0)      199641          14
test_identity_function_overhead_benchmark[cvm-trust_input=True]             439.9999 (1.36)     15,960.0002 (2.81)       505.0075 (1.41)     135.9068 (1.79)       500.9999 (1.41)     29.9999 (2.46)      136;793    1,980.1684 (0.71)      32064           1
test_identity_function_overhead_benchmark[cvm_nogc-trust_input=False]     1,312.0000 (4.06)      8,876.9998 (1.56)     1,446.4540 (4.04)     179.0074 (2.36)     1,442.0000 (4.06)     49.9999 (4.09)      114;493      691.3459 (0.25)      36521           1
test_identity_function_overhead_benchmark[cvm-trust_input=False]          1,382.0002 (4.28)      8,705.9998 (1.53)     1,525.5059 (4.26)     198.9647 (2.63)     1,512.9999 (4.26)     50.9999 (4.18)       42;142      655.5202 (0.23)       7106           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

```
</details> 


***

I also removed the strict zip, because passing `strict=False` is more than 100ns slower than not specifying it at all. Check out with this snippet:

```python
a = tuple(range(10))
b = tuple(range(9))
%timeit tuple(zip(a, b, strict=False))
%timeit tuple(zip(a, b))
587 ns ± 8.29 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
458 ns ± 17.4 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

</details>

Closes #1332 
Does the same as and closes #1976 
Closes #1647

<!-- readthedocs-preview pytensor start -->
----
📚 Documentation preview 📚: https://pytensor--1351.org.readthedocs.build/en/1351/

<!-- readthedocs-preview pytensor end -->

ricardoV94 committed:
Remove deprecated function overhead and reduce overhead

ricardoV94 committed:
Remove DefaultOp

ricardoV94 committed:
Fix Eigvalsh handling of b=None

Armavica commented:
Did you mean to add the `DefaultOp` and `Eigvalsh` as part of this PR?
No problem if that's the case but I am trying to understand if they are related to the function overhead

ricardoV94 commented:
Those Ops relied on special casing of inputs being None at runtime (but not typed as such at definition time) which required a general extra `if input is not None` guard before `type.filter` all the time.

Armavica commented on doc/library/compile/io.rst:
```diff
@@ -73,52 +72,25 @@ The ``inputs`` argument to ``pytensor.function`` is a list, containing the ``Var
 
-Value: initial and default values
----------------------------------
+Update
```
I like this mechanism compared to what was described before. But does this mean that the `update` parameter of `In` is deprecated or not useful anymore?

ricardoV94 replied:
it's used internally, and still how shared variable updates are determined. It's just restricted to shared variables though

Armavica commented on doc/library/compile/io.rst:
```diff
@@ -225,56 +138,6 @@ Every element of the inputs list will be upgraded to an In instance if necessary
 
 - a tuple ``(name, r)`` will be ``In(r, name=name)``
 
-- a tuple ``(r, val)`` will be ``In(r, value=value, autoname=True)``
-
-- a tuple ``((r,up), val)`` will be ``In(r, value=value, update=up, autoname=True)``
-
-- a tuple ``(name, r, val)`` will be ``In(r, name=name, value=value)``
-
-- a tuple ``(name, (r,up), val)`` will be ``In(r, name=name, value=val, update=up, autoname=True)``
```
Good riddance :D 
I had no clue that those existed and they seem more confusing than useful to me

Armavica commented on pytensor/compile/function/types.py:
```diff
@@ -474,41 +467,20 @@ def __init__(
-
+        self._finder = finder = {}
+        self._inv_finder = inv_finder = {}
+        self._named_inputs = named_inputs = []
```
Are these performance optimizations or typos?

ricardoV94 replied:
Not typos. The underscore is just so they don't show as easily to users of a compiled function. The alias not saved in self is so we can nicely reference it in the rest of this code block as we build them

Armavica commented on pytensor/compile/function/types.py:
```diff
@@ -594,46 +505,43 @@ def __contains__(self, item):
 
         # Function.__call__ is responsible for updating the inputs, unless the vm promises to do it itself
-        self.update_input_storage: tuple[int, Container] = ()
+        self._update_input_storage: tuple[int, Container] = ()
```
This is a `tuple[tuple[int, Container], ...]`

ricardoV94 replied:
Yeah, I always get tripped by this

Armavica commented on pytensor/compile/function/types.py:
```diff
@@ -908,136 +915,30 @@ def __call__(self, *args, output_subset=None, **kwargs):
+                storage_data[0] = arg
+            if kwargs:  # for speed, skip the items for empty kwargs
+                for k, arg in kwargs.items():
+                    self._finder[k].storage[0] = arg
```
This gained an indentation level, I can't decide if that was on purpose

ricardoV94 replied:
it's now duplicated in the two branches, so we don't need to check if self.trust_input again. The use in the else branch is encapsulated in the helper

Armavica commented:
I understood most of it (not all) and it looks mostly okay to me, except in a couple of places where I have doubts

jessegrabowski commented on pytensor/compile/function/types.py:
```diff
@@ -330,23 +330,35 @@ class Function:
     """
 
+    __slots__ = (
+        # Created inside __init__
```
remove

jessegrabowski commented on pytensor/compile/function/types.py:
```diff
@@ -391,20 +401,11 @@ def __init__(
             tuple elements are used only by Kits, which are deprecated.
```
What are Kits? Should we clean up this stuff as part of this PR?

ricardoV94 replied:
I have no idea

ricardoV94 replied:
Something supposedly removed in 2016. Good riddance: https://github.com/Theano/Theano/pull/4619

jessegrabowski commented on pytensor/compile/function/types.py:
```diff
@@ -1561,6 +1428,20 @@ def __init__(
+            for i in inputs
+        ):
+            raise ValueError(
+                "Inputs with default values were deprecated. Use `functools.partial` instead."
```
```suggestion
                "Inputs with default values are deprecated. Use `functools.partial` instead."
```

jessegrabowski commented on pytensor/compile/function/types.py:
```diff
@@ -1561,6 +1428,20 @@ def __init__(
         # Wrap them in In or Out instances if needed.
         inputs = [self.wrap_in(i) for i in inputs]
+
+        # Remove this after a while
```
Mark with TODO or something so we can find this? It's not a super useful comment.

jessegrabowski commented on pytensor/compile/function/types.py:
```diff
@@ -861,26 +768,126 @@ def checkSV(sv_ori, sv_rpl):
-    def __call__(self, *args, output_subset=None, **kwargs):
+        # Check if inputs are missing, or if inputs were set more than once, or
+        # if we tried to provide inputs that are supposed to be implicit.
+        for arg_container in input_storage:
```
it seems like we should be able to do this book-keeping along the way so we don't need another loop over the inputs. But it's just a one-time cost so whatever.

ricardoV94 replied:
This mixes info collected from args and kwargs, so we can only know if something is missing or provided multiple times if there are no kwargs?

jessegrabowski commented on tests/compile/function/test_types.py:
```diff
@@ -522,107 +414,111 @@ def test_shared_state0(self):
         f(1, 2)
-        assert f[s] == 2
-        assert g[s] == 2
+        assert f._finder[s].value == 2
```
we could implement `__getitem__` for function as `return self._finder[idx]` ?

ricardoV94 replied:
I don't want `__getitem__` for function, it's not at all obvious why there should be one

jessegrabowski commented on tests/compile/function/test_types.py:
```diff
@@ -805,78 +685,11 @@ def test_output_dictionary(self):
-
-        assert result[0] == 47.0
-        assert result[1] == 420.0
+        with pytest.raises(ValueError, match="output_keys was deprecated"):
```
is it even worth testing this?

ricardoV94 replied:
Doesn't hurt much

jessegrabowski commented on tests/compile/function/test_types.py:
```diff
@@ -1286,7 +1038,7 @@ def exc_message(e):
     def test_pickle_class_with_functions(self):
         blah = SomethingToPickle()
-        assert blah.f2.container[blah.s].storage is blah.f1.container[blah.s].storage
+        assert blah.f2._finder[blah.s].storage is blah.f1._finder[blah.s].storage
```
does the variable have to be called blah?

ricardoV94 replied:
clearly, blue would be too confusing

jessegrabowski reviewed (approved):
-1.2k, you love to see it

ricardoV94 committed:
Merge pfunc into function

Remove pfunc as a separate entry point. function() now directly calls
construct_function_ins_and_outs + FunctionMaker, eliminating an
unnecessary layer of indirection.

Also:
- Inline _pfunc_param_to_in as a local closure (param_to_in), drop
  unused strict parameter
- Fix stale docstring in construct_function_ins_and_outs

Remove redundant tests between TestFunction (ex-TestPfunc) and
TestFunctionIn:
- test_param_strict (duplicate of test_in_strict)
- test_param_mutable (duplicate of test_in_mutable)
- test_param_allow_downcast_int (duplicate of test_in_allow_downcast_int)
- test_param_allow_downcast_floatX (duplicate of test_in_allow_downcast_floatX)
- test_param_allow_downcast_vector_floatX (duplicate of test_in_allow_downcast_vector_floatX)
- test_explicit_shared_input (duplicate of test_no_shared_as_input)

ricardoV94 committed:
Remove compile/function package

Replace with flat modules:
- compile/maker.py: function(), FunctionMaker
- compile/executor.py: Function class, pickle support,
AliasedMemoryError
- compile/aliasing.py: Supervisor, alias/view/deepcopy utilities
- compile/rebuild.py: rebuild_collect_shared,
construct_function_ins_and_outs

Also moves get_info_on_inputs into Function as a static method and
inlines fgraph_updated_vars at its only call site.

ricardoV94 committed:
Move debug/profiling modules into compile.debug subpackage

- compile/debugmode.py -> compile/debug/debugmode.py
- compile/nanguardmode.py -> compile/debug/nanguardmode.py
- compile/monitormode.py -> compile/debug/monitormode.py
- compile/profiling.py -> compile/debug/profiling.py
- Move function_dump into compile/debug/dump.py

ricardoV94 committed:
Remove orig_function wrapper

Replace orig_function usage in Scan with direct mode.function_maker()
call. The profiling that orig_function did is already handled by
FunctionMaker itself.

ricardoV94 committed:
Move std_fgraph into FunctionMaker.create_fgraph

Subsume std_fgraph as a staticmethod on FunctionMaker and inline
_optcheck_fgraph in DebugMode._Maker where it was the only caller.

Also fixes a bug where the clone parameter was computed but not
passed to FunctionGraph.__init__.

ricardoV94 committed:
Rename function_builder to function_class in FunctionMaker

ricardoV94 committed:
Inline and remove iter_over_pairs helper

Only used internally in construct_function_ins_and_outs, replace with
direct isinstance(updates, dict) checks at the two call sites.

ricardoV94 commented:
mypy failures look unrelated...

ricardoV94 commented:
Second round, scope creep so original reviews are pretty much moot

ricardoV94 commented on doc/library/compile/io.rst:
```diff
@@ -225,56 +138,6 @@ Every element of the inputs list will be upgraded to an In instance if necessary
 
 - a tuple ``(name, r)`` will be ``In(r, name=name)``
```
TODO : Remove this as well

ricardoV94 committed:
Fix type hint
